astrokat/noisediode.py

```diff
@@ -9,257 +9,419 @@
-    kat : Session kat container-like object
+    kat : session kat container-like object
         Container for accessing KATCP resources allocated to schedule block.
-    session : katcorelib.CaptureSession-like object
-    nd_setup: dict
+    nd_setup : dict
         Noise diode pattern setup, with keys:
-            'antennas':  options are `ants`, or 'm062',...
-            'cycle_len': the cycle length [sec], - must be less than 20 sec for L-band
+            'antennas':  options are 'all', or 'm062', or ....,
+            'cycle_len': the cycle length [sec],
+                           - must be less than 20 sec for L-band,
+            etc., etc.
     lead_time : float, optional
-                Lead time before digitisers pattern is set [sec]
-
+        Lead time before digitisers pattern is set [sec]
     """
-    nd_antennas = nd_setup["antennas"]  # selected antennas for nd pattern
-    cycle_length = nd_setup["cycle_len"]  # nd pattern length [sec]
-    on_fraction = nd_setup["on_frac"]  # on fraction of pattern length [%]
-    msg = (
-        "Repeat noise diode pattern every {} sec, with {} sec on and apply pattern to"
-        " {}".format(cycle_length, float(cycle_length) * float(on_fraction), nd_antennas)
-    )
-    user_logger.info(msg)
+    lead_time = _eval_lead_time_(lead_time)
 
-    if not kat.dry_run:
-        if kat.sensor.sub_band.get_value() == "l" and float(cycle_length) > 20.0:
-            msg = "Maximum cycle length of L-band is 20 seconds"
-            raise RuntimeError(msg)
-        # Set noise diode period to multiple of correlator integration time.
-        dump_period = session.cbf.correlator.sensor.int_time.get_value()
-        user_logger.warning(
-            "Correlator integration time {} [sec]".format(1.0 / dump_period)
-        )
-        cycle_length = int(cycle_length / dump_period) * dump_period
-        msg = "Set noise diode period to multiple of correlator dump period:"
-        msg += " cycle length = {} [sec]".format(cycle_length)
-        user_logger.warning(msg)
+    # nd pattern length [sec]
+    _MAX_CYCLE_LEN = _get_max_cycle_len(kat)
+    if float(nd_setup['cycle_len']) > _MAX_CYCLE_LEN:
+        msg = 'Maximum cycle length is {} seconds'.format(_MAX_CYCLE_LEN)
+        raise RuntimeError(msg)
+
+    user_logger.trace('TRACE: max cycle len {}'
+                      .format(_MAX_CYCLE_LEN))
 
     # Try to trigger noise diodes on all antennas in array simultaneously.
     # - use integer second boundary as that is most likely be an exact
     #   time that DMC can execute at, and also fit a unix epoch time
     #   into a double precision float accurately
     # - add a default lead time to ensure enough time for all digitisers
     #   to be set up
-    timestamp = np.ceil(time.time() + lead_time)
-    msg = (
-        "Request: Set noise diode pattern to activate at {} "
-        "(includes {} sec lead time)".format(timestamp, lead_time)
-    )
+    start_time = _set_time_(lead_time)
```

Comment:
@ajoubertza you are correct, the comment should actually only read
```
# Try to trigger noise diodes on specified antennas in array simultaneously.
# - add a default lead time to ensure enough time for all digitisers
#   to be set up
```
I updated that to be in line with the functionality

The `ceil` was removed since it enforced an assumption. I thought it better to send the request time (float as is) and wait for the digitiser to report the actual timestamp applied.
That way it should be more future proof ...

Once the ND functionality is sound and gets used actively, it will be easier to refine the noise diode behaviour and instructions from observation scripts, which can then be implemented and evaluated against the actual observations.
Currently, this is a little difficult for me to predict

Reply:
The choice of integer second boundary was made after discussions with Marc.  My understanding was that the digitisers could accurately execute a command on a second boundary, giving that they are synchronised with a pulse-per-second (PPS) clock.

You are right that we can refine things after detailed evaluation of real observations.  My guesstimate is that not truncating the fractional part will introduce a little more variance into the timing, as the fractional parts of the times are essentially uniform random numbers.  Worst case, the digitiser will be half a clock cycle period away from the time we actually asked for.

Whether we use ceil or not, using the actual timestamps reported by the digitiser is critical, so that is good.